File: 63. Unique Paths II.py
# 63. Unique Paths 2


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        rows, cols = len(obstacleGrid), len(obstacleGrid[0])
        dp = [0] * cols
        dp[-1] = 1
        for r in range(rows - 1, -1, -1):
            for c in range(cols - 1, -1, -1):
                if obstacleGrid[r][c] == 1:
                    dp[c] = 0
                else:
                    if c < cols - 1:
                        dp[c] = dp[c] + dp[c + 1]
        return dp[0]

File: 63. Unique Paths II.cpp
// 63. Unique Paths II


class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int rows = obstacleGrid.size(), cols = obstacleGrid[0].size();
        vector<long> dp (cols, 0);
        dp[cols-1] = 1;

        for (int i = rows-1; i >= 0; i--) {
            for (int j = cols-1; j >= 0; j--) {
                if (obstacleGrid[i][j] == 1) {
                    dp[j] = 0;
                } else if (j < cols -1) {
                    dp[j] = dp[j] + dp[j+1];
                }
            }
        }
        return dp[0];
    }
};